src/message.c

```diff
@@ -134,32 +134,40 @@ static int ldbus_message_get_type(lua_State *L) {
 
 static int ldbus_message_iter_init(lua_State *L) {
 	DBusMessage *message = check_DBusMessage(L, 1);
-	DBusMessageIter *iter;
+	lDBusMessageIter *iter;
 	if (lua_gettop(L) == 1) {
 		push_DBusMessageIter(L);
 	} else {
 		lua_settop(L, 2);
+		unref_ldbus_message_iter(L, 2);
 	}
 	iter = luaL_checkudata(L, 2, DBUS_MESSAGE_ITER_METATABLE);
```

Comment:
I'd prefer that we didn't call `touserdata` in `unref_ldbus_message_iter`, but instead just had it take the iter.
So you get here e.g.

```suggestion
	lDBusMessageIter *iter;
	if (lua_gettop(L) == 1) {
		push_DBusMessageIter(L);
		iter = luaL_checkudata(L, 2, DBUS_MESSAGE_ITER_METATABLE); // or `lua_touserdata` as we know the type is correct already
	} else {
		lua_settop(L, 2);
		iter = luaL_checkudata(L, 2, DBUS_MESSAGE_ITER_METATABLE);
		unref_ldbus_message_iter(iter);
	}
```

As an additional improvement we could make `push_DBusMessageIter` return the pointer directly; but perhaps leave that for another day.

Reply:
Thanks, I applied the style in the new HEAD.